src/gtclang/Unittest/SirGenerator.h

```diff
@@ -0,0 +1,204 @@
+std::shared_ptr<dawn::AssignmentExpr> assign(const std::shared_ptr<dawn::Expr>& left,
+                                             const std::shared_ptr<dawn::Expr>& right, const char *op="=");
+std::shared_ptr<dawn::TernaryOperator> ternop(const std::shared_ptr<dawn::Expr>& cond,
+                                              const std::shared_ptr<dawn::Expr>& left,
+                                              const std::shared_ptr<dawn::Expr>& right);
+std::shared_ptr<dawn::FunCallExpr> fcall(const std::string& callee);
+std::shared_ptr<dawn::StencilFunCallExpr> sfcall(const std::string& calee);
+std::shared_ptr<dawn::StencilFunArgExpr> sfarg(int direction, int offset, int argumentIndex);
+std::shared_ptr<dawn::VarAccessExpr> var(const std::string& name,
+                                         std::shared_ptr<dawn::Expr> index = nullptr);
+std::shared_ptr<dawn::FieldAccessExpr>
+field(const std::string& name, dawn::Array3i offset = dawn::Array3i{{0, 0, 0}},
+      dawn::Array3i argumentMap = dawn::Array3i{{-1, -1, -1}},
+      dawn::Array3i argumentOffset = dawn::Array3i{{0, 0, 0}}, bool negateOffset = false);
+std::shared_ptr<dawn::LiteralAccessExpr> lit(const std::string& value,
+                                             dawn::BuiltinTypeID builtinType = dawn::BuiltinTypeID::Integer);
+/// @}
+
+/// @brief Simple Visitor that prints Statements with their respective Types
+/// @ingroup unittest
+class PrintAllExpressionTypes : public dawn::ASTVisitorForwarding {
+public:
+  virtual void visit(const std::shared_ptr<dawn::BlockStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::ExprStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::ReturnStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::VarDeclStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::VerticalRegionDeclStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::StencilCallDeclStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::BoundaryConditionDeclStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::IfStmt>& stmt);
+  virtual void visit(const std::shared_ptr<dawn::UnaryOperator>& expr);
+  virtual void visit(const std::shared_ptr<dawn::BinaryOperator>& expr);
+  virtual void visit(const std::shared_ptr<dawn::AssignmentExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::TernaryOperator>& expr);
+  virtual void visit(const std::shared_ptr<dawn::FunCallExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::StencilFunCallExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::StencilFunArgExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::VarAccessExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::FieldAccessExpr>& expr);
+  virtual void visit(const std::shared_ptr<dawn::LiteralAccessExpr>& expr);
+};
+
+/// @brief helper to finds all the field to register for auto-generating stencil codes
+class FieldFinder : public dawn::ASTVisitorForwarding {
```

Comment:
I would just expose a sinlge C function called `findFieldsInStmt` or something that takes an AST and returns a vector of the found fields. The visitor class can be moved into a private translation unit and called by the `findFieldsInStmt`. This enhances overall readability and makes it easier (as well as more generic) to use this functionality. I would even consider moving this into Dawn (maybe add it to https://github.com/MeteoSwiss-APN/dawn/blob/master/src/dawn/SIR/ASTUtil.h)